### PR TITLE
[Abacus] Fix normalize

### PR DESCRIPTION
**Pull request by hvdijk**
# Overview

[Abacus] Fix normalize

# Reason for change

The implementation had bugs. The testing tried to paper over those bugs, unsuccessfully, but should not be trying that.

# Description of change

* Unlike fast_normalize(), normalize() must not allow overflow or underflow of internal calculations. Use scaling to avoid issues.
* The handling of infinity and NaN was not right. We were ignoring the sign of elements in flushing to zero, and we were wrongly letting infinity take priority over NaN: normalize(<Infinity, NaN>) should return NaN, not <1, 0>.
* Fix incorrect ULP values being reported because of confusion over half-as-float and half-as-int. calcHalfPrecisionULP takes a half-as-int value, but was passed a half-as-float value.
* Improve error reporting to print floating point values in decimal form too, not just their representation in hex form.
* We were trying to flag valid calls to normalize() as undefined to account for implementation bugs.

# Anything else we should know?

*If there's any other relevant information we should know that may help us in
understanding and verifying your patch, please include it here.*

# Checklist

* Read and follow the project [Code of Conduct](https://github.com/codeplaysoftware/oneapi-construction-kit/blob/main/CODE_OF_CONDUCT.md).
* Make sure the project builds successfully with your changes.
* Run relevant testing locally to avoid regressions.
* Run [clang-format-17](https://clang.llvm.org/docs/ClangFormat.html) on all modified code.

### CODE REVIEW

File: modules/compiler/builtins/abacus/generate/abacus_detail_geometric/abacus_detail_geometric.in [hunk 1]
```diff
@@ -223,56 +223,105 @@ struct fast_length_untyped_helper<T, true> {
...
-  const IntType isInf = __abacus_isinf(p);
-
-  // Broadcasts scalar abacus_int to vector type since `__abacus_any()` returns
-  // scalar 1 if any vector lane has a sign bit set.
-  const IntType isAnyInf = __abacus_any(isInf) == 1;
+T normalize(T p) {
+  constexpr size_t num_elements = TypeTraits<T>::num_elements;
+  static_assert(num_elements >= 1 && num_elements <= 16);
+  if constexpr (num_elements == 1) {  // NOLINT
+    return p > 0 ? 1 : p < 0 ? -1 : p;
```
Comment (coldav) [inline modules/compiler/builtins/abacus/generate/abacus_detail_geometric/abacus_detail_geometric.in:257]: I find it hard to scan this, can we add a few brackets?

File: modules/compiler/builtins/abacus/generate/abacus_detail_geometric/abacus_detail_geometric.in [hunk 2]
```diff
@@ -223,56 +223,105 @@ struct fast_length_untyped_helper<T, true> {
...
+    if (max_mag == 0) {
+      // If all elements were zero, handle this specially by returning the
+      // vector unmodified.
+      return p;
+    }
 
-  // Flush positive INF to 1.
-  const IntType isPosInf = isInf & ~__abacus_signbit(p);
-  infsFlushed = __abacus_select(infsFlushed, T(1.0), isPosInf);
+    if (max_mag == 0) {
```
Comment (coldav) [inline modules/compiler/builtins/abacus/generate/abacus_detail_geometric/abacus_detail_geometric.in:280]: is this not the same as L274?

Comment (hvdijk) [reply modules/compiler/builtins/abacus/generate/abacus_detail_geometric/abacus_detail_geometric.in:280]: It is, thanks, I think I messed up the conflict resolution after a rebase, will remove.

File: modules/compiler/builtins/abacus/generate/abacus_detail_geometric/abacus_detail_geometric.in (hunk 1 above)
Comment (hvdijk) [reply modules/compiler/builtins/abacus/generate/abacus_detail_geometric/abacus_detail_geometric.in:257]: I'll change this to use `if` statements which probably help even more.

**Review by coldav — APPROVED** [review]

**Pull request merged**